integration/ztoc_test.go

```diff
@@ -145,3 +179,248 @@ func ztocExistChecker(t *testing.T, listOutputLines []string, img testImageIndex
+				} else if err == nil {
+					t.Fatal("failed to return error")
+				}
+			})
+		}
+	}
+
+}
+
+func TestSociZtocGetFile(t *testing.T) {
+	t.Parallel()
+	sh, done := newSnapshotterBaseShell(t)
+	defer done()
+	rebootContainerd(t, sh, "", "")
+
+	testImages := prepareSociIndices(t, sh)
+
+	var (
+		tempOutputStream = "test.txt"
+		randomFile       = string(testutil.RandomByteData(10))
+		randomZtocDigest = testutil.RandomDigest()
+	)
+
+	getRandomFilePathsWithinZtoc := func(t *testing.T, ztocDigest string, numFiles int) []string {
+		rand.Seed(time.Now().UnixNano())
+		var (
+			zinfo               Info
+			regPaths, randPaths []string
+		)
+		output := sh.O("soci", "ztoc", "info", ztocDigest)
+		json.Unmarshal(output, &zinfo)
+		for _, file := range zinfo.Files {
+			if file.Type == "reg" {
+				regPaths = append(regPaths, file.Filename)
+			}
+		}
+		for i := 0; i < numFiles; i++ {
+			randPaths = append(randPaths, regPaths[rand.Intn(len(regPaths))])
+		}
+		return randPaths
+	}
+
+	verifyOutputStream := func(t *testing.T, contents, output []byte) error {
+		d := cmp.Diff(contents, output)
+		if d == "" {
+			return nil
+		}
+		return fmt.Errorf("expected %s got %s", contents, output)
+	}
+
+	for _, img := range testImages {
+		img := img
+		ztocDigest := img.ztocDigests[0]
+		var layerDigest string
+		sociIndex, _ := sociIndexFromDigest(sh, img.sociIndexDigest)
+
+		for _, blob := range sociIndex.Blobs {
+			if blob.Digest.String() == ztocDigest {
+				layerDigest = blob.Annotations[soci.IndexAnnotationImageLayerDigest]
+				break
+			}
+		}
+		layerContents := sh.O("cat", filepath.Join(containerdBlobStorePath, trimSha256Prefix(layerDigest)))
+		files := getRandomFilePathsWithinZtoc(t, ztocDigest, 10)
+
+		testCases := []struct {
+			name        string
+			cmd         []string
+			toStdout    bool
+			expectedErr bool
+		}{
+			{
+				name:        "Ztoc that does not exist",
+				cmd:         []string{"soci", "ztoc", "get-file", randomZtocDigest, randomFile},
+				toStdout:    true,
+				expectedErr: true,
+			},
+			{
+				name:        "Ztoc exists but file does not exist",
+				cmd:         []string{"soci", "ztoc", "get-file", ztocDigest, randomFile},
+				toStdout:    true,
+				expectedErr: true,
+			},
+			{
+				name:        "Ztoc and each file exists, file contents redirected to stdout",
+				cmd:         []string{"soci", "ztoc", "get-file", ztocDigest},
+				toStdout:    true,
+				expectedErr: false,
+			},
+			{
+				name:        "Ztoc and each file exists, file contents redirected to output file",
+				cmd:         []string{"soci", "ztoc", "get-file", "-o", tempOutputStream, ztocDigest},
+				toStdout:    false,
+				expectedErr: false,
+			},
+		}
+		for _, tt := range testCases {
+			t.Run(tt.name, func(t *testing.T) {
+				if !tt.expectedErr {
+					for _, f := range files {
+						cmd := append(tt.cmd, f)
+						output, err := sh.OLog(cmd...)
+						if err != nil {
+							t.Fatalf("failed to return file contents")
+						}
```

Comment:
we should print the err if not nil: `t.Fatalf("failed to return file contents: %v", err)`